File: python/144.py
# 二叉树的前序遍历
# Binary Tree Preorder Traversal

from common import List, Optional, TreeNode, create_binary_tree


class Solution:
    def preorder_traversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def visit(node: Optional[TreeNode], res: List[int]):
            if node is None:
                return
            res.append(node.val)
            visit(node.left, res)
            visit(node.right, res)

        visit(root, res)

        return res


if __name__ == "__main__":
    test_cases = [
        ([1, None, 2, 3],),
        ([1, 2, 3, 4, 5, None, 8, None, None, 6, 7, 9],),
    ]
    sol = Solution()
    for case in test_cases:
        root = create_binary_tree(case[0])
        print(sol.preorder_traversal(root))

File: cpp/144.cpp
// 二叉树的前序遍历
// Binary Tree Preorder Traversal

#include "common.h"

class Solution
{
public:
    std::vector<int> preorderTraversal(TreeNode *root)
    {
        std::vector<int> result;
        visit(root, result);
        return result;
    }

    void visit(TreeNode *node, std::vector<int> &result)
    {
        if (node == nullptr)
            return;
        result.push_back(node->val);
        visit(node->left, result);
        visit(node->right, result);
    }
};

int main()
{
    std::vector<std::vector<int>> test_cases = {
        {1, NULL_VALUE, 2, 3},
        {1, 2, 3, 4, 5, NULL_VALUE, 8, NULL_VALUE, NULL_VALUE, 6, 7, 9},
    };

    Solution sol;
    for (const auto &test_case : test_cases)
    {
        TreeNode *root = create_binary_tree(test_case);
        std::vector<int> result = sol.preorderTraversal(root);
        print_vector(result);
    }
}
